File: udon_decompiler/analysis/transform/passes/detect_exit_points.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Optional

from udon_decompiler.analysis.ir.nodes import (
    IRBlock,
    IRBlockContainer,
    IRFunction,
    IRIf,
    IRJump,
    IRLeave,
    IRStatement,
)
from udon_decompiler.analysis.transform.pass_base import (
    IILTransform,
    ILTransformContext,
)


class _ExitSentinel:
    pass


EXIT_NOT_YET_DETERMINED = _ExitSentinel()
NO_EXIT = _ExitSentinel()


@dataclass
class _ExitOccurrence:
    statement: IRStatement
    replace: Callable[[IRStatement], None]


class DetectExitPoints(IILTransform):
    """Port of ILSpy DetectExitPoints, adapted to the current Python IR."""

    def __init__(self, can_introduce_exit_for_return: bool) -> None:
        self.can_introduce_exit_for_return = can_introduce_exit_for_return

        self._function_body: Optional[IRBlockContainer] = None
        self._current_container: Optional[IRBlockContainer] = None
        self._current_exit: object = NO_EXIT
        self._potential_exits: Optional[list[_ExitOccurrence]] = None

        self._descendant_block_cache: dict[int, set[IRBlock]] = {}

    def run(self, function: IRFunction, context: ILTransformContext) -> None:
        self._function_body = function.body
        self._current_container = None
        self._current_exit = NO_EXIT
        self._potential_exits = None
        self._descendant_block_cache = {}

        self._visit_container(
            container=function.body,
            parent_block=None,
            parent_index=None,
        )

    @staticmethod
    def compatible_exit_instruction(exit1: object, exit2: object) -> bool:
        if type(exit1) is not type(exit2):
            return False

        if isinstance(exit1, IRJump) and isinstance(exit2, IRJump):
            return exit1.target is exit2.target

        if isinstance(exit1, IRLeave) and isinstance(exit2, IRLeave):
            return exit1.target_container is exit2.target_container

        return False

    def _visit_container(
        self,
        container: IRBlockContainer,
        parent_block: Optional[IRBlock],
        parent_index: Optional[int],
    ) -> None:
        old_exit = self._current_exit
        old_container = self._current_container
        old_potential_exits = self._potential_exits

        this_exit = self._get_exit_after_statement(parent_block, parent_index)

        self._current_exit = this_exit
        self._current_container = container
        self._potential_exits = (
            [] if this_exit is EXIT_NOT_YET_DETERMINED else None
        )

        for block in container.blocks:
            self._visit_block(block)

        if (
            this_exit is EXIT_NOT_YET_DETERMINED
            and self._potential_exits is not None
            and self._potential_exits
            and parent_block is not None
            and parent_index is not None
        ):
            chosen_exit = self._choose_exit(self._potential_exits)
            for occurrence in self._potential_exits:
                if self.compatible_exit_instruction(chosen_exit, occurrence.statement):
                    occurrence.replace(IRLeave(target_container=container))

            parent_block.statements.insert(
                parent_index + 1,
                self._clone_exit_statement(chosen_exit),
            )

        self._current_exit = old_exit
        self._current_container = old_container
        self._potential_exits = old_potential_exits

    def _visit_block(self, block: IRBlock) -> None:
        index = 0
        while index < len(block.statements):
            statement = block.statements[index]

            def replace_current(new_statement: IRStatement, i: int = index) -> None:
                block.statements[i] = new_statement

            self._visit_statement(
                statement=statement,
                replace_statement=replace_current,
                parent_block=block,
                parent_index=index,
            )
            index += 1

    def _visit_statement(
        self,
        statement: IRStatement,
        replace_statement: Callable[[IRStatement], None],
        parent_block: Optional[IRBlock],
        parent_index: Optional[int],
    ) -> None:
        if isinstance(statement, IRBlockContainer):
            self._visit_container(statement, parent_block, parent_index)
            return

        if isinstance(statement, IRIf):
            self._visit_statement(
                statement=statement.true_statement,
                replace_statement=lambda new_stmt: setattr(
                    statement,
                    "true_statement",
                    new_stmt,
                ),
                parent_block=None,
                parent_index=None,
            )
            if statement.false_statement is not None:
                self._visit_statement(
                    statement=statement.false_statement,
                    replace_statement=lambda new_stmt: setattr(
                        statement,
                        "false_statement",
                        new_stmt,
                    ),
                    parent_block=None,
                    parent_index=None,
                )
            return

        if isinstance(statement, IRJump):
            if self._current_container is None:
                return

            if not self._is_block_descendant_of_container(
                statement.target,
                self._current_container,
            ):
                self._handle_exit(statement, replace_statement)
            return

        if isinstance(statement, IRLeave):
            self._handle_exit(statement, replace_statement)

    def _handle_exit(
        self,
        statement: IRStatement,
        replace_statement: Callable[[IRStatement], None],
    ) -> None:
        if self._current_container is None:
            return

        if self._current_exit is EXIT_NOT_YET_DETERMINED:
            if (
                self._potential_exits is not None
                and self._can_introduce_as_exit(statement)
            ):
                self._potential_exits.append(
                    _ExitOccurrence(statement=statement, replace=replace_statement)
                )
            return

        if self.compatible_exit_instruction(statement, self._current_exit):
            replace_statement(IRLeave(target_container=self._current_container))

    def _can_introduce_as_exit(self, statement: IRStatement) -> bool:
        if self._current_container is None:
            return False

        if self._container_has_leave_to_self(self._current_container):
            return False

        if isinstance(statement, IRLeave) and self._is_function_return_leave(statement):
            return self.can_introduce_exit_for_return

        return isinstance(statement, (IRJump, IRLeave))

    def _container_has_leave_to_self(self, container: IRBlockContainer) -> bool:
        for block in container.blocks:
            for statement in block.statements:
                if self._statement_has_leave_target(statement, container):
                    return True
        return False

    def _statement_has_leave_target(
        self,
        statement: IRStatement,
        target_container: IRBlockContainer,
    ) -> bool:
        if isinstance(statement, IRLeave):
            return statement.target_container is target_container

        if isinstance(statement, IRIf):
            if self._statement_has_leave_target(
                statement.true_statement,
                target_container,
            ):
                return True
            if statement.false_statement is not None and (
                self._statement_has_leave_target(
                    statement.false_statement,
                    target_container,
                )
            ):
                return True
            return False

        if isinstance(statement, IRBlockContainer):
            for block in statement.blocks:
                for nested in block.statements:
                    if self._statement_has_leave_target(nested, target_container):
                        return True

        return False

    def _is_function_return_leave(self, statement: IRLeave) -> bool:
        if self._function_body is None:
            return False
        return statement.target_container is self._function_body

    def _get_exit_after_statement(
        self,
        parent_block: Optional[IRBlock],
        parent_index: Optional[int],
    ) -> object:
        if parent_block is None or parent_index is None:
            return NO_EXIT

        next_index = parent_index + 1
        if next_index < len(parent_block.statements):
            return parent_block.statements[next_index]

        return EXIT_NOT_YET_DETERMINED

    def _clone_exit_statement(self, statement: object) -> IRStatement:
        if isinstance(statement, IRJump):
            return IRJump(target=statement.target)
        if isinstance(statement, IRLeave):
            return IRLeave(target_container=statement.target_container)
        raise TypeError("Exit statement must be IRJump or IRLeave")

    def _choose_exit(self, exits: list[_ExitOccurrence]) -> IRStatement:
        first = exits[0].statement
        if isinstance(first, IRLeave) and self._is_function_return_leave(first):
            for occurrence in exits[1:]:
                candidate = occurrence.statement
                if not (
                    isinstance(candidate, IRLeave)
                    and self._is_function_return_leave(candidate)
                ):
                    return candidate
        return first

    def _is_block_descendant_of_container(
        self,
        block: IRBlock,
        container: IRBlockContainer,
    ) -> bool:
        cache_key = id(container)
        descendants = self._descendant_block_cache.get(cache_key)
        if descendants is None:
            descendants = self._collect_descendant_blocks(container)
            self._descendant_block_cache[cache_key] = descendants
        return block in descendants

    def _collect_descendant_blocks(self, container: IRBlockContainer) -> set[IRBlock]:
        descendants: set[IRBlock] = set()

        def visit_container(current: IRBlockContainer) -> None:
            for block in current.blocks:
                descendants.add(block)
                for statement in block.statements:
                    if isinstance(statement, IRBlockContainer):
                        visit_container(statement)
                    elif isinstance(statement, IRIf):
                        visit_statement(statement.true_statement)
                        if statement.false_statement is not None:
                            visit_statement(statement.false_statement)

        def visit_statement(statement: IRStatement) -> None:
            if isinstance(statement, IRBlockContainer):
                visit_container(statement)
                return
            if isinstance(statement, IRIf):
                visit_statement(statement.true_statement)
                if statement.false_statement is not None:
                    visit_statement(statement.false_statement)

        visit_container(container)
        return descendants

File: src/decompiler/transform/passes/detect_exit_points.rs
use std::collections::{HashMap, HashSet};

use crate::decompiler::Result;
use crate::decompiler::ir::{
    IrBlock, IrBlockContainer, IrFunction, IrHighLevelDoWhile, IrHighLevelSwitch, IrHighLevelWhile,
    IrIf, IrJump, IrLeave, IrStatement,
};
use crate::decompiler::transform::pass_base::{ITransform, TransformContext};

#[derive(Debug, Clone)]
enum CurrentExit {
    NotYetDetermined,
    NoExit,
    Exit(Box<IrStatement>),
}

#[derive(Debug, Clone)]
enum StatementPathSegment {
    Block(usize),
    Statement(usize),
    IfTrue,
    IfFalse,
}

#[derive(Debug, Clone)]
struct ExitOccurrence {
    statement: IrStatement,
    path: Vec<StatementPathSegment>,
}

pub struct DetectExitPoints {
    pub can_introduce_exit_for_return: bool,
}

impl DetectExitPoints {
    pub fn new(can_introduce_exit_for_return: bool) -> Self {
        Self {
            can_introduce_exit_for_return,
        }
    }

    pub fn compatible_exit_instruction(a: &IrStatement, b: &IrStatement) -> bool {
        match (a, b) {
            (
                IrStatement::Jump(IrJump { target_address: ta }),
                IrStatement::Jump(IrJump { target_address: tb }),
            ) => ta == tb,
            (
                IrStatement::Leave(IrLeave {
                    target_container_id: ta,
                }),
                IrStatement::Leave(IrLeave {
                    target_container_id: tb,
                }),
            ) => ta == tb,
            _ => false,
        }
    }
}

impl ITransform for DetectExitPoints {
    fn run(
        &self,
        function: &mut IrFunction,
        _context: &mut TransformContext<'_, '_>,
    ) -> Result<()> {
        let mut state = DetectState::new(function.body.id, self.can_introduce_exit_for_return);
        let _ = state.visit_container(&mut function.body, CurrentExit::NoExit);
        Ok(())
    }
}

struct DetectState {
    function_body_id: u32,
    can_introduce_exit_for_return: bool,

    current_container_id: Option<u32>,
    current_exit: CurrentExit,
    potential_exits: Option<Vec<ExitOccurrence>>,

    descendant_block_cache: HashMap<u32, HashSet<u32>>,
    self_leave_cache: HashMap<u32, bool>,
}

impl DetectState {
    fn new(function_body_id: u32, can_introduce_exit_for_return: bool) -> Self {
        Self {
            function_body_id,
            can_introduce_exit_for_return,
            current_container_id: None,
            current_exit: CurrentExit::NoExit,
            potential_exits: None,
            descendant_block_cache: HashMap::new(),
            self_leave_cache: HashMap::new(),
        }
    }

    fn visit_container(
        &mut self,
        container: &mut IrBlockContainer,
        current_exit: CurrentExit,
    ) -> Option<IrStatement> {
        let old_exit = self.current_exit.clone();
        let old_container_id = self.current_container_id;
        let old_potential_exits = self.potential_exits.take();

        self.current_exit = current_exit;
        self.current_container_id = Some(container.id);
        let has_self_leave = statement_has_leave_target_container(container, container.id);
        self.self_leave_cache.insert(container.id, has_self_leave);
        self.potential_exits = if matches!(self.current_exit, CurrentExit::NotYetDetermined) {
            Some(Vec::new())
        } else {
            None
        };

        let descendant_blocks = self.descendant_blocks(container);

        for (block_index, block) in container.blocks.iter_mut().enumerate() {
            self.visit_block(block, block_index, container.id, &descendant_blocks);
        }

        let should_introduce = matches!(self.current_exit, CurrentExit::NotYetDetermined)
            && self.potential_exits.as_ref().is_some_and(|x| !x.is_empty());

        let introduced_exit = if should_introduce {
            let chosen_exit = self.choose_exit(self.potential_exits.as_ref().expect("present"));

            if let Some(exits) = self.potential_exits.as_ref() {
                for occurrence in exits {
                    if Self::compatible_ref(&chosen_exit, &occurrence.statement) {
                        let replaced = replace_statement_at_path(
                            container,
                            &occurrence.path,
                            IrStatement::Leave(IrLeave {
                                target_container_id: container.id,
                            }),
                        );
                        debug_assert!(
                            replaced,
                            "invalid statement path recorded in DetectExitPoints"
                        );
                    }
                }
            }

            Some(self.clone_exit_statement(&chosen_exit))
        } else {
            None
        };

        self.current_exit = old_exit;
        self.current_container_id = old_container_id;
        self.potential_exits = old_potential_exits;
        introduced_exit
    }

    fn visit_block(
        &mut self,
        block: &mut IrBlock,
        block_index: usize,
        current_container_id: u32,
        descendant_blocks: &HashSet<u32>,
    ) {
        let mut index = 0usize;
        while index < block.statements.len() {
            let next_exit = self.get_exit_after_statement(Some(block), Some(index));
            let mut path = vec![
                StatementPathSegment::Block(block_index),
                StatementPathSegment::Statement(index),
            ];
            let inserted_exit = {
                let statement = &mut block.statements[index];
                self.visit_statement(
                    statement,
                    current_container_id,
                    descendant_blocks,
                    Some(next_exit),
                    &mut path,
                )
            };
            if let Some(inserted_exit) = inserted_exit {
                let insert_at = (index + 1).min(block.statements.len());
                block.statements.insert(insert_at, inserted_exit);
            }
            index += 1;
        }
    }

    fn visit_statement(
        &mut self,
        statement: &mut IrStatement,
        current_container_id: u32,
        descendant_blocks: &HashSet<u32>,
        current_exit: Option<CurrentExit>,
        path: &mut Vec<StatementPathSegment>,
    ) -> Option<IrStatement> {
        match statement {
            IrStatement::BlockContainer(container) => {
                self.visit_container(container, current_exit.unwrap_or(CurrentExit::NoExit))
            }
            IrStatement::If(IrIf {
                true_statement,
                false_statement,
                ..
            }) => {
                path.push(StatementPathSegment::IfTrue);
                let _ = self.visit_statement(
                    true_statement,
                    current_container_id,
                    descendant_blocks,
                    None,
                    path,
                );
                path.pop();
                if let Some(false_statement) = false_statement.as_mut() {
                    path.push(StatementPathSegment::IfFalse);
                    let _ = self.visit_statement(
                        false_statement,
                        current_container_id,
                        descendant_blocks,
                        None,
                        path,
                    );
                    path.pop();
                }
                None
            }
            IrStatement::Block(block) => {
                for (index, nested) in block.statements.iter_mut().enumerate() {
                    path.push(StatementPathSegment::Statement(index));
                    let _ = self.visit_statement(
                        nested,
                        current_container_id,
                        descendant_blocks,
                        None,
                        path,
                    );
                    path.pop();
                }
                None
            }
            IrStatement::Jump(IrJump { target_address }) => {
                if !descendant_blocks.contains(target_address) {
                    self.handle_exit(statement, current_container_id, path);
                }
                None
            }
            IrStatement::Leave(_) => {
                self.handle_exit(statement, current_container_id, path);
                None
            }
            _ => None,
        }
    }

    fn handle_exit(
        &mut self,
        statement: &mut IrStatement,
        current_container_id: u32,
        path: &[StatementPathSegment],
    ) {
        match &self.current_exit {
            CurrentExit::NotYetDetermined => {
                if self.can_introduce_as_exit(statement)
                    && let Some(potential_exits) = self.potential_exits.as_mut()
                {
                    potential_exits.push(ExitOccurrence {
                        statement: statement.clone(),
                        path: path.to_vec(),
                    });
                }
            }
            CurrentExit::Exit(exit) => {
                if Self::compatible_ref(statement, exit.as_ref()) {
                    *statement = IrStatement::Leave(IrLeave {
                        target_container_id: current_container_id,
                    });
                }
            }
            CurrentExit::NoExit => {}
        }
    }

    fn can_introduce_as_exit(&self, statement: &IrStatement) -> bool {
        let Some(container_id) = self.current_container_id else {
            return false;
        };

        if self.container_has_leave_to_self(container_id) {
            return false;
        }

        match statement {
            IrStatement::Leave(IrLeave {
                target_container_id,
            }) if *target_container_id == self.function_body_id => {
                self.can_introduce_exit_for_return
            }
            IrStatement::Jump(_) | IrStatement::Leave(_) => true,
            _ => false,
        }
    }

    fn container_has_leave_to_self(&self, container_id: u32) -> bool {
        self.self_leave_cache
            .get(&container_id)
            .copied()
            .unwrap_or(false)
    }

    fn get_exit_after_statement(
        &self,
        parent_block: Option<&IrBlock>,
        parent_index: Option<usize>,
    ) -> CurrentExit {
        let (Some(parent_block), Some(parent_index)) = (parent_block, parent_index) else {
            return CurrentExit::NoExit;
        };

        let next_index = parent_index + 1;
        if next_index < parent_block.statements.len() {
            return CurrentExit::Exit(Box::new(parent_block.statements[next_index].clone()));
        }

        CurrentExit::NotYetDetermined
    }

    fn clone_exit_statement(&self, statement: &IrStatement) -> IrStatement {
        match statement {
            IrStatement::Jump(IrJump { target_address }) => IrStatement::Jump(IrJump {
                target_address: *target_address,
            }),
            IrStatement::Leave(IrLeave {
                target_container_id,
            }) => IrStatement::Leave(IrLeave {
                target_container_id: *target_container_id,
            }),
            _ => statement.clone(),
        }
    }

    fn choose_exit(&self, exits: &[ExitOccurrence]) -> IrStatement {
        let first = exits[0].statement.clone();
        if is_function_return_leave(&first, self.function_body_id) {
            for occurrence in exits.iter().skip(1) {
                if !is_function_return_leave(&occurrence.statement, self.function_body_id) {
                    return occurrence.statement.clone();
                }
            }
        }
        first
    }

    fn compatible_ref(a: &IrStatement, b: &IrStatement) -> bool {
        DetectExitPoints::compatible_exit_instruction(a, b)
    }

    fn descendant_blocks(&mut self, container: &IrBlockContainer) -> HashSet<u32> {
        if let Some(cached) = self.descendant_block_cache.get(&container.id) {
            return cached.clone();
        }

        let mut out = HashSet::<u32>::new();
        collect_descendant_block_addresses(container, &mut out);
        self.descendant_block_cache
            .insert(container.id, out.clone());
        out
    }
}

fn replace_statement_at_path(
    container: &mut IrBlockContainer,
    path: &[StatementPathSegment],
    replacement: IrStatement,
) -> bool {
    let Some((StatementPathSegment::Block(block_index), rest)) = path.split_first() else {
        return false;
    };
    let Some(block) = container.blocks.get_mut(*block_index) else {
        return false;
    };
    replace_statement_in_block(block, rest, replacement)
}

fn replace_statement_in_block(
    block: &mut IrBlock,
    path: &[StatementPathSegment],
    replacement: IrStatement,
) -> bool {
    let Some((StatementPathSegment::Statement(statement_index), rest)) = path.split_first() else {
        return false;
    };
    let Some(statement) = block.statements.get_mut(*statement_index) else {
        return false;
    };
    replace_statement_in_statement(statement, rest, replacement)
}

fn replace_statement_in_statement(
    statement: &mut IrStatement,
    path: &[StatementPathSegment],
    replacement: IrStatement,
) -> bool {
    let Some((segment, rest)) = path.split_first() else {
        *statement = replacement;
        return true;
    };

    match (segment, statement) {
        (StatementPathSegment::IfTrue, IrStatement::If(if_stmt)) => {
            replace_statement_in_statement(if_stmt.true_statement.as_mut(), rest, replacement)
        }
        (StatementPathSegment::IfFalse, IrStatement::If(if_stmt)) => if_stmt
            .false_statement
            .as_mut()
            .is_some_and(|false_statement| {
                replace_statement_in_statement(false_statement.as_mut(), rest, replacement)
            }),
        (StatementPathSegment::Statement(statement_index), IrStatement::Block(block)) => block
            .statements
            .get_mut(*statement_index)
            .is_some_and(|nested| replace_statement_in_statement(nested, rest, replacement)),
        _ => false,
    }
}

fn is_function_return_leave(statement: &IrStatement, function_body_id: u32) -> bool {
    matches!(
        statement,
        IrStatement::Leave(IrLeave { target_container_id }) if *target_container_id == function_body_id
    )
}

fn collect_descendant_block_addresses(container: &IrBlockContainer, out: &mut HashSet<u32>) {
    for block in &container.blocks {
        out.insert(block.start_address);
        for statement in &block.statements {
            collect_descendants_from_statement(statement, out);
        }
    }
}

fn collect_descendants_from_statement(statement: &IrStatement, out: &mut HashSet<u32>) {
    match statement {
        IrStatement::BlockContainer(container) => {
            collect_descendant_block_addresses(container, out)
        }
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            collect_descendants_from_statement(true_statement, out);
            if let Some(false_statement) = false_statement.as_ref() {
                collect_descendants_from_statement(false_statement, out);
            }
        }
        IrStatement::Block(block) => {
            out.insert(block.start_address);
            for nested in &block.statements {
                collect_descendants_from_statement(nested, out);
            }
        }
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            for section in sections {
                out.insert(section.body.start_address);
                for nested in &section.body.statements {
                    collect_descendants_from_statement(nested, out);
                }
            }
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            collect_descendant_block_addresses(body, out)
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            collect_descendant_block_addresses(body, out)
        }
        _ => {}
    }
}

fn statement_has_leave_target_container(
    container: &IrBlockContainer,
    target_container_id: u32,
) -> bool {
    for block in &container.blocks {
        for statement in &block.statements {
            if statement_has_leave_target(statement, target_container_id) {
                return true;
            }
        }
    }
    false
}

fn statement_has_leave_target(statement: &IrStatement, target_container_id: u32) -> bool {
    match statement {
        IrStatement::Leave(IrLeave {
            target_container_id: id,
        }) => *id == target_container_id,
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            statement_has_leave_target(true_statement, target_container_id)
                || false_statement
                    .as_ref()
                    .is_some_and(|x| statement_has_leave_target(x, target_container_id))
        }
        IrStatement::Block(block) => block
            .statements
            .iter()
            .any(|x| statement_has_leave_target(x, target_container_id)),
        IrStatement::BlockContainer(container) => {
            statement_has_leave_target_container(container, target_container_id)
        }
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            sections.iter().any(|section| {
                section
                    .body
                    .statements
                    .iter()
                    .any(|x| statement_has_leave_target(x, target_container_id))
            })
        }
        IrStatement::HighLevelWhile(IrHighLevelWhile { body, .. }) => {
            statement_has_leave_target_container(body, target_container_id)
        }
        IrStatement::HighLevelDoWhile(IrHighLevelDoWhile { body, .. }) => {
            statement_has_leave_target_container(body, target_container_id)
        }
        _ => false,
    }
}
